test/shared/src/main/scala/zio/test/TestAspect.scala

```diff
@@ -162,6 +161,33 @@ object TestAspect extends TimeoutVariants {
         effect *> test
     }
 
+  /**
+   * An aspect that runs each test on a separate fiber and prints a fiber dump
+   * if the test fails or has not terminated within the specified duration.
+   */
+  def diagnose(duration: Duration): TestAspectAtLeastR[Live] =
+    new TestAspectAtLeastR[Live] {
+      def some[R <: Live, E](predicate: String => Boolean, spec: ZSpec[R, E]): ZSpec[R, E] = {
+        def diagnose[R <: Live, E](
+          label: String,
+          test: ZIO[R, TestFailure[E], TestSuccess]
+        ): ZIO[R, TestFailure[E], TestSuccess] =
+          test.fork.flatMap { fiber =>
+            Live.live(clock.sleep(duration)) *> fiber.poll.flatMap {
```

Comment:
What happens if test completes before `duration` elapsed?

Reply:
Good point. We should race the test with the sleep fiber. Then if the test completes first we can just print the fiber dump if the test failed and interrupt the sleep fiber. If the sleep fiber completes first we can print the fiber dump and then join back the test fiber. Thank you!